### ISSUE TITLE
Detection of ECB mode, false negatives

### ISSUE THREAD

**Issue by john-aws** [FuLoRi/Analyze-LastPassVaultGUI#8]
I wonder if the CBC detection in Analyze-LastPassVaultGUI is 100% correct. It seems to be testing solely for the regex pattern: `'^!'` but according to [What the LastPass CLI tells us about LastPass Design](https://systemoverlord.com/2015/09/16/what-the-lastpass-cli-tells-us-about-lastpass-design/) blog post, the LastPass CLI's AES cipher mode detection [here](https://github.com/lastpass/lastpass-cli/blob/master/cipher.c#L160) is more specific:

```c
if (len >= 33 && len % 16 == 1 && ciphertext[0] == '!') {
    // it's CBC mode
} else {
    // it's ECB mode
}
```

The rationale for the more complex test is:

> if they just checked the first character of the ciphertext, then 1 out of 256 ECB-mode encrypted ciphertexts would match that [ed: _so would be incorrectly identified as CBC_]. Since ECB mode ciphertexts are multiples of the block length (as are CBC ciphertexts), checking for the length to have one extra character (len % 16 == 1) rules out these extra cases.

**Comment by FuLoRi** [comment]
Hi @john-aws!

Have you tested the script with that modification to see if it works as expected? I would prefer to use it (assuming it works) to be more accurate in the detection. The regex it's currently using is the one Steve suggested. 

Thanks!

**Comment by john-aws** [comment]
After a quick test, it's clear that the `len` that the C code is testing is not the length of the ciphertext so it's not as simple as substituting `$result.$encryptedValue.Length` for `len` above.

So, I'll need to do some testing here and I'll update when I find out more. I think this is an edge case but it would be nice to remove the, admittedly small, chance of false negatives.

**Comment by FuLoRi** [comment]
I was in a rush and didn't look closely at the code or I would've realized that it's C and not PowerShell. 😆

I have low confidence in my ability to understand C, so I asked ChatGPT and it said:

> The first check is to verify that the length of the ciphertext (stored in the variable "len") is greater than or equal to 33 and that the length is also divisible by 16 with a remainder of 1. This is likely a specific requirement for the format of the ciphertext when it is encrypted using CBC mode.

I asked it to convert the C code to PowerShell and it gave me this:
```powershell
if((($len -gt 33) -and ($len % 16 -eq 1)) -and ($ciphertext[0] -eq '!')) {
    # it's CBC mode
} else {
    # it's ECB mode
}
```

If ChatGPT is right, then it seems to me that we should be able to substitute `$len` with `$result.$encryptedValue.Length` and `$ciphertext` with `$result.$encryptedValue`, like this:
```powershell
if((($result.$encryptedValue.Length -gt 33) -and ($result.$encryptedValue.Length % 16 -eq 1)) -and ($result.$encryptedValue[0] -eq '!')) {
    # it's CBC mode
} else {
    # it's ECB mode
}
```

Am I missing anything? Let me know what you think.

Thank you!

**Comment by john-aws** [comment]
Hi Rob, the C to PowerShell transpilation of that test is certainly valid code but the length being tested in the two situations (`len` in the C code vs. `$result.$encryptedValue.Length` in the PS1 code) don't refer to the same thing. Using that simple transpilation against my vault, for example, results in all accounts being marked as `WARNING: Encrypted with ECB!`

Also, my understanding is that LastPass encryption yields the following, depending on which AES mode was used:

```
ECB => base64-encoded ciphertext
CBC => ! + base64-encoded IV + | + base64-encoded ciphertext
```

So that should be what you see in `$result.$encryptedValue`. I don't have any account credentials encrypted with ECB mode so I don't have actual data to validate this assumption and test against.

Anyhow, if LastPass yields a base64-encoded value (alone) for ECB, it can't begin with the `!` (bang) character because that's not in the base64 alphabet `[A-Za-z0-9+/]`. Let me do a little more digging. I'm going to close this issue for now but will re-open if needed.
